### PR TITLE
Window grouping

### PR DESCRIPTION
**Pull request by mussel**
This set of commits implements the window grouping functionality described in #439, where multiple windows of a given application share the same IconButton. To choose between windows, the user clicks in the application's button and a Popover appears with a list of the opened windows. 

The application menu (the one that appears by right-clicking the IconButton) operates on the "Current Window", i.e. the last window of an application that was active.

When a window needs attention, the applet will initiate the "urgency animation" in both the application's button and also the window's row in the aforementioned dropdown list. This allows the user to identify exactly which window is requesting the attention.

Lastly, I also managed to make Spotify behave nicely when pinned. As far as I can tell, in the master version of the tasklist applet it is possible to pin Spofity. However, when you open it, instead of using the PinnedIconButton already available, the applet was creating a new IconButton for Spotify. In this new version, once Spotify sets its window class, the applet is able to query its DesktopAppInfo and use it to find the PinnedIconButton already associated with Spotify and then use it, instead of creating a new button.

Now regarding the Pull Request itself.

 I Tried to cleanup the commit as best as I could and for the most part I think it is ok. However, I ended up making extensive changes to both Buttons.vala and IconTasklistApplet.vala files, due to refactoring and the architectural changes necessary to implement the new functionality.  I personally think the amount of changes these files is not ideal, but it would be non-trivial to make it tidier. Take a look and if you think I should split the changes into more commits I can give it a try.

### CODE REVIEW

File: panel/applets/icon-tasklist/Application.vala [hunk 1]
```diff
@@ -0,0 +1,283 @@
+public class Application: AttentionStatusListener, IconChangeListener, ScreenProvider{
```
Comment (ikeydoherty) [inline panel/applets/icon-tasklist/Application.vala]: Missing license headers

File: panel/applets/icon-tasklist/Application.vala [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+        current_appwin = appwin;
+        window_list.AddWindow (appwin);
+
+        // Setup event listeners
+        appwin.SetAttentionStatusListener (this);
+        appwin.SetIconChangeListener (this);
+        // configures the window minimization animation
+        UpdateWindowMinimizeAnimation (window);
+
+        if (GetWindowNum () == 1){
```
Comment (ikeydoherty) [inline panel/applets/icon-tasklist/Application.vala]: This Windows style isn't a valid coding style in Budgie

Should be: get_window_num, C style

File: panel/applets/icon-tasklist/Application.vala [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+    public GLib.DesktopAppInfo? GetInfo (){
+        return app_info;
+    }
+
+    /**
+     * Returns a reference to the most recently opened window or null if there
+     * is no window opened
+     */
+    public Wnck.Window? GetCurrentWindow (){
```
Comment (ikeydoherty) [inline panel/applets/icon-tasklist/Application.vala]: Spaces between function name and parentheses shouldn't be there

File: panel/applets/icon-tasklist/Application.vala [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+        Wnck.Window? window = null;
+        if (current_appwin != null){
```
Comment (ikeydoherty) [inline panel/applets/icon-tasklist/Application.vala]: Space after parentheses should be here

File: panel/applets/icon-tasklist/Application.vala (hunk 2 above)
Comment (mussel) [inline panel/applets/icon-tasklist/Application.vala]: If understand this correctly, this "C style" coding style should be used for all function names, right?

File: panel/applets/icon-tasklist/Application.vala (hunk 1 above)
Comment (mussel) [inline panel/applets/icon-tasklist/Application.vala]: I will use the same license headers as in the original .vala files. But I have one question. In the files that I created, I'm supposed to put my name and email in the Copyright line, right?

Sorry, I've never done this before.

Comment (ikeydoherty) [inline panel/applets/icon-tasklist/Application.vala]: Oh definitely copyright it to you! It's your hard work, after all. Gotta make sure you're credited for it :D

File: panel/applets/icon-tasklist/Application.vala (hunk 2 above)
Comment (ikeydoherty) [inline panel/applets/icon-tasklist/Application.vala]: Yeah :)

**Comment by QORTEC** [comment]
Thank you @mussel, this is one of those "core" features that I've been waiting for a little while now... 

I was just wondering, will it be possible to disable window grouping via option menu or config file...?
Personally I would never want to disabled it, I can imagine individuals with workflows who may dislike window grouping.

**Comment by mussel** [comment]
@QORTEC disabling the functionality is easy, but I do not know how to expose this as a config. Perhaps, @ikeydoherty could give me some pointers on how to do it.

**Comment by mussel** [comment]
@ikeydoherty  I've made the changes you asked me. See if there is anything else that should be changed.

As I was saying in #439 , this PR implements the basic functionality for window grouping. But there are still aesthetic and usability changes that might have to be implemented (see suggestions in #439). So @ikeydoherty , is it ok to get this PR merged first and then work those changes or should I implement everything first before merging the PR?

**Comment by mussel** [comment]
I did not know which would be the best way to highlight the active window of an application. So I implemented the highlighting in two different ways:

highlighting1_two_elements
highlighting_three_elements

and

highlighting2_two_elements
highlighting2_three_elements


Personally I would prefer a variation of the first implementation, but I am not an UX guys. What do you guys think.
The code for the highlighting is on the branches: 
https://github.com/mussel/budgie-desktop/tree/window_grouping_actwin_highlight 
and 
https://github.com/mussel/budgie-desktop/tree/window_grouping_actwin_highlight2

**Comment by markbaas** [comment]
Nice PR, but I guess this will needs to be ported to C as well.

**Comment by ikeydoherty** [comment]
Which I'll help with, obviously.

File: panel/applets/icon-tasklist/UrgencyAnimation.vala [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+            return false;
+        }
+
+        return true;
+    }
+
+    private bool draw(Cairo.Context cr)
+    {
+        if (!is_animation_in_progress()) {
+            message ("Intercepting draw signal but animation is not in progress");
```
Comment (ikeydoherty) [inline panel/applets/icon-tasklist/UrgencyAnimation.vala]: Why is this here?

File: panel/applets/icon-tasklist/WindowListItem.vala [hunk 6]
```diff
@@ -0,0 +1,138 @@
+/*
+ * This file is part of budgie-desktop
+ *
+ * Copyright (C) 2016 Fernando Mussel <fernandomussel91@gmail.com>
```
Comment (ikeydoherty) [inline panel/applets/icon-tasklist/WindowListItem.vala]: This isn't right. I see my own code in this file. i.e. gracefully_die.

File: panel/applets/icon-tasklist/WindowListView.vala [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    public void remove_window(ApplicationWindow appwin)
+    {
+        var list_row = window_row_map [appwin];
+
+        // Handles the removal, with animation, of WindowListItem from the ListBox
+        list_row.gracefully_die(window_listbox);
+        window_row_map.remove(appwin);
+        // hold item to prevent its destruction and the possible resize of the
```
Comment (ikeydoherty) [inline panel/applets/icon-tasklist/WindowListView.vala]: Why are you not using multiline comments ?

File: panel/applets/icon-tasklist/WindowManager.vala [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * Copyright (C) 2016 Fernando Mussel <fernandomussel91@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ */
+
+public interface WindowManager
```
Comment (ikeydoherty) [inline panel/applets/icon-tasklist/WindowManager.vala]: Why all of this abstraction? How many WindowManager implementations do you foresee..?

File: panel/applets/icon-tasklist/WindowMenu.vala [hunk 9]
```diff
@@ -0,0 +1,161 @@
+/*
+ * This file is part of budgie-desktop
+ *
+ * Copyright (C) 2016 Fernando Mussel <fernandomussel91@gmail.com>
```
Comment (ikeydoherty) [inline panel/applets/icon-tasklist/WindowMenu.vala]: Again you use my code here, you cannot remove my copyright

File: panel/panel.vala [hunk 10]
```diff
@@ -952,22 +952,22 @@ public class Panel : Budgie.Toplevel
         }
         this.expanded = expanded;
         if (!expanded) {
-            if (wm_proxy != null) {
+            /*if (wm_proxy != null) {
```
Comment (ikeydoherty) [inline panel/panel.vala]: What is this? You're disabling core functionality of the panel

File: panel/applets/icon-tasklist/WindowMenu.vala [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+            }
+            sep = new Gtk.SeparatorMenuItem();
+            menu.append(sep);
+            sep.show_all();
+            foreach (var action in actions) {
+                var display_name = app_info.get_action_name(action);
+                var item = new Gtk.MenuItem.with_label(display_name);
+                item.set_data("__aname", action);
+                item.activate.connect(()=> {
+                    string? act = item.get_data("__aname");
```
Comment (ikeydoherty) [inline panel/applets/icon-tasklist/WindowMenu.vala]: Would be nice if we could move away from lambdas tbh, Vala lambdas are a heap of shite at best and damn near impossible to debug

File: panel/applets/icon-tasklist/WindowListView.vala [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    /**
+     * Add new WindowListItem associated with \p appwin
+     */
+    public void add_window(ApplicationWindow appwin)
+    {
+        WindowListItem win_row = add_window_list_item(appwin);
+
+        window_row_map [appwin] = win_row;
```
Comment (ikeydoherty) [inline panel/applets/icon-tasklist/WindowListView.vala]: No spaces on `[ ]`

**Comment by ikeydoherty** [comment]
I wanna add that I'm still positive about the work here, it's just making sure we integrate it in a way that is maintainable, as it's most likely going to be me porting it. Thus I need to understand it fully :)

File: panel/applets/icon-tasklist/WindowListItem.vala (hunk 6 above)
Comment (mussel) [inline panel/applets/icon-tasklist/WindowListItem.vala]: Sorry, this was not intentional. The code is new but I wanted to animate the ListBoxRow removal, so I used your code for that. I will be sure to change this and be more attentive to it in the future.

File: panel/applets/icon-tasklist/WindowListView.vala (hunk 7 above)
Comment (mussel) [inline panel/applets/icon-tasklist/WindowListView.vala]: It is a habit thing. The multi-line comments make it impossible to comment out sections of the code, which I do from time to time for debugging.
I can change that to the multi-line ones if you prefer.

File: panel/applets/icon-tasklist/UrgencyAnimation.vala (hunk 5 above)
Comment (ikeydoherty) [inline panel/applets/icon-tasklist/UrgencyAnimation.vala]: Oh I actually meant the print message :p

File: panel/applets/icon-tasklist/WindowManager.vala (hunk 8 above)
Comment (mussel) [inline panel/applets/icon-tasklist/WindowManager.vala]: yeah, you have a point. I used an interface to avoid coupling the classes to a concrete implementation (IconTasklistApplet in this case). But at the moment, this is an unnecessary complexity.

**Comment by mussel** [comment]
Thanks for taking the time to review this, I really appreciate it. =]

So on to your main concerns:
1. **Removing the copyright**: Stupid mistake on my part. It was not intentional and won't happen again.
2. **Undoing core functionality**: I will implement a different solution (which I summarized in my comment response) that does not require disabling this new functionality.
3. **Heavy levels of abstraction**:  Thanks for pointing out the obvious regarding the WindowManager interface. But I also wanted to know if you though other interfaces were unnecessary, such as the listener interfaces that I use in the ApplicationWindow class.

PS: If you think it would be better to port to this to C I would be willing to help as well. I would probably need some guidance, but could give you a hand.

File: panel/applets/icon-tasklist/UrgencyAnimation.vala (hunk 5 above)
Comment (mussel) [inline panel/applets/icon-tasklist/UrgencyAnimation.vala]: My bad ;)
Once the animation ends I unregister the callback to prevent the draw function to be endlessly called for no reason. So I put the message there to let us know if callback was been called when it was not supposed to, to help debug this issue if it were to happen.

File: panel/applets/icon-tasklist/WindowListItem.vala (hunk 6 above)
Comment (mussel) [inline panel/applets/icon-tasklist/WindowListItem.vala]: So, for this file that includes code from both of us, I would also need to credit authorship to both of us. So is there a specific standard that the project is using? For instance, the oldest copyright comes first:

Copyright (C) 2014-2016 Ikey Doherty <ikey@solus-project.com>
Copyright (C) 2016 Fernando Mussel <fernandomussel91@gmail.com>

File: panel/applets/icon-tasklist/WindowManager.vala (hunk 8 above)
Comment (mussel) [inline panel/applets/icon-tasklist/WindowManager.vala]: One extra thing regarding the WindowManager. Although the interface might be unnecessary, I am in favor of keeping at least some of the functionality it exposes concentrated in IconTasklistApplet to avoid code duplication and keeping things more organized. For instance:
 * *toggle_window*: the function is a bit convoluted and is used in three different places; two times in Button.vala and once in WindowListItem.vala

* *toggle_window_list*: used only once in Button.vala, but it requires a reference to popover_manager. We could give a reference to the manager to every button, but IconTasklistApplet already has to listen to changes to the popover_manager and unregister the popovers belonging to applications that are closed completely (0 windows left). So I thought it would be better to just let IconTasklistApplet handle display the popover.
* *close_window*: used in only two places. It is simple enough that we could probably just remove it.
* *query_window*: I created it to avoid having to pass yet another object, in this case an AppSystem instance, to Application. I don't know. Maybe we should keep this function.

File: panel/panel.vala (hunk 10 above)
Comment (mussel) [inline panel/panel.vala]: Actually, one thing that I just thought. What about modifying the function *show_popover* of the PopoverManager, to allows us to specify if we want the previous active window to be made active again? We could add a  boolean argument to select the desired behavior. If we were to call:
```vala
popover_manager.show_popover(btn, false);
```
Once the popover associated with *btn* was closed, nothing would happen. 
To avoid breaking compatibility in the vala interface, we could set this boolean value to default to true (to make the previous active window active again). For the C interface, this would require exposing a function with a different name.

Would such a solution be fine with you?

Comment (ikeydoherty) [inline panel/panel.vala]: So my thoughts here is that, legitimately, our use of `GtkPopover` in budgie is the #1 thing breaking everything. I think the real solution is to stop `budgie-panel` using the window expansion system and switch to real top level windows

**Comment by ikeydoherty** [comment]
Closing as Budgie is now being made modular - when the dust settles please make a new PR against the new components :)
